security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,70 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal"
+	"hash"
+)
+
+type S2AHalfConnection struct {
+	aeadCrypter   s2aAeadCrypter
+	expander      hkdfExpander
+	seqCounter    counter
+	trafficSecret []byte
+	nonce         []byte
+}
+
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := NewCiphersuite(ciphersuite)
+	if cs.expectedTrafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bits, given: %v", cs.expectedTrafficSecretSize()*8, trafficSecret)
+	}
+
+	hc := S2AHalfConnection{expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
+	key, err := hc.deriveSecret(cs.hashFunction(), trafficSecret, []byte("tls13 key"))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, \"tls13 key\") failed with error: %v", trafficSecret, err)
+	}
+	hc.nonce, err = hc.deriveSecret(cs.hashFunction(), trafficSecret, []byte("tls13 iv"))
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.deriveSecret(h, %v, \"tls13 iv\") failed with error: %v", trafficSecret, err)
+	}
+	hc.aeadCrypter, err = cs.aeadCrypter(key)
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("cs.aeadCrypter(%v) failed with error: %v", key, err)
+	}
+	return hc, nil
+}
+
+func (hc *S2AHalfConnection) Encrypt(dst, plaintext, aad []byte) ([]byte, error) {
+	// TODO(rnkim): Implement this.
+	panic("Encrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) Decrypt(dst, ciphertext, aad []byte) ([]byte, error) {
+	// TODO(rnkim): Implement this.
+	panic("Decrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) UpdateKey(key []byte) error {
```

Comment:
This function should not have a parameter. It has to derive the new key from the traffic secret.